Add support for setting update interval dynamically from WPJMCOM

fjorgemota opened the pull request:
Alternative to #2561

### Changes proposed in this Pull Request

* Changes the Promoted Jobs Status Handle to run the check hourly  (but it WILL NOT run hourly unless the server specifies to do so)
* Split the update interval setting into two different options: CRON and WEBHOOK. The first is considered when the WP-CRON is triggering the request, and the other is considering in every other call (usually coming out from the webhook).
* Every time the status handler fetches an update from the server, it will look for the `X-WPJM-Webhook-Interval` and `X-WPJM-Cron-Interval` headers and will update the value of the respective option accordingly (or delete it, if the header's value is `false`)
* The system considers always the last execution time (whether it comes from the cron or the webhook) when running the update, to not run very often
* From WPJM side, this PR is effectively identical to what's already implemented on the feature branch. Except that now we have the power to change the options at will from WPJM. :D 

### Testing instructions

**NOTE**: Consider that "running the webhook" is the same as running the command ` curl -X POST "http://your.wpjm.instance/?rest_route=/wpjm-internal/v1/promoted-jobs/refresh-status"`

**TIP**: You can use `wp option list --search="job_manager_promoted_jobs_*"` to see the options related to Promoted Jobs. :) 

On a clean WP installation running this branch and running a WPJMCOM instance based on 581-gh-Automattic/wpjobmanager.com.

1. Make sure the options `job_manager_promoted_jobs_cron_interval` and `job_manager_promoted_jobs_webhook_interval` aren't set
2. Try running the webhook using  and verify that it doesn't make any requests (you can check that by verifying if the option `job_manager_promoted_jobs_status_update_last_execution` doesn't exist)
3. Promote a job
4. Make sure the options `job_manager_promoted_jobs_cron_interval` and `job_manager_promoted_jobs_webhook_interval` are set and are numbers (the cron_interval should be set to 43200, and the webhook_interval to 300). Make sure the `job_manager_promoted_jobs_status_update_last_execution` option still doesn't exist (if it exists, delete it)
5. Try running the webhook now and make sure the option `job_manager_promoted_jobs_status_update_last_execution` is set
7. Delete the Promoted Site (and their Promoted Jobs) from WPJMCOM side
8. Delete the `job_manager_promoted_jobs_status_update_last_execution` (or set it to `0`, it is effectively the same)
9. Try running the webhook again and verify if the options  `job_manager_promoted_jobs_cron_interval` and `job_manager_promoted_jobs_webhook_interval` are deleted now

Sorry for the complicated testing instructions, by the way. :)

fjorgemota committed:
Update the update interval for cron/webhook based on headers in the response

fjorgemota committed:
Consider the update interval based on whether the code is running inside a cron or not

fjorgemota committed:
Update intervals for cron/webhooks before redirecting to WPJMCOM

fjorgemota committed:
Reduce update interval for the cron job when redirecting to WPJMCOM

fjorgemota committed:
Reduce interval for the cron to fetch promoted jobs updates

The check will run more often, but it will be still limited by CRON_INTERVAL_OPTION_KEY. If we want to run it only twice a day, it is fine. If we want to run it more often, it is fine, too! :)

fjorgemota committed:
Disable autoload for the option for saving intervals for Promoted Jobs Status Handler

This optimizes performance very slightly, as WP will have one (or two) option less to load on every request. :D

fjorgemota committed:
Delete option from the database if the header is "false"

This effectively removes an option from the database, and makes the code uses its default "0" when checking for this option

fjorgemota committed:
Do not update option if it already exists before redirecting user to WPJMCOM

To avoid risking overwriting whatever value the server served just because the user decided to promote a new job

fjorgemota committed:
Also do not load LAST_EXECUTION_OPTION_KEY on every request

Because it is simply not needed on every where

fjorgemota committed:
Always update interval on response from WPJM, independent if it is an error or not

So we can catch conditions where the Promoted Site doesn't exist anymore, for instance

fjorgemota committed:
Update check to verify if header exist on the response

Because apparently array_key_exists doesn't work as expected with the custom class that WP uses to implement case insensitive dictionaries

fjorgemota commented on includes/class-wp-job-manager.php:
```diff
@@ -251,7 +251,7 @@ public static function maybe_schedule_cron_jobs() {
 		}
 		if ( ! wp_next_scheduled( WP_Job_Manager_Promoted_Jobs_Status_Handler::CRON_HOOK ) ) {
-			wp_schedule_event( time(), 'twicedaily', WP_Job_Manager_Promoted_Jobs_Status_Handler::CRON_HOOK );
+			wp_schedule_event( time(), 'hourly', WP_Job_Manager_Promoted_Jobs_Status_Handler::CRON_HOOK );
```
It is important to note that, although this change is a bit weird at first, it is only done to increase the number of checks on the fallback per day.

We're still limited to running only twice a day on the internal check that the Promoted Jobs Status Handler verify. BUT we now run this check much more often, which allows us to increase the number of checks we do if we want on WPJMCOM side.

fjorgemota commented on includes/admin/class-wp-job-manager-promoted-jobs-admin.php:
```diff
@@ -209,7 +209,8 @@ public function handle_promote_job() {
 
-		update_option( WP_Job_Manager_Promoted_Jobs_Status_Handler::USED_PROMOTED_JOBS_OPTION_KEY, true );
+		add_option( WP_Job_Manager_Promoted_Jobs_Status_Handler::CRON_INTERVAL_OPTION_KEY, 12 * HOUR_IN_SECONDS, '', false );
+		add_option( WP_Job_Manager_Promoted_Jobs_Status_Handler::WEBHOOK_INTERVAL_OPTION_KEY, 5 * MINUTE_IN_SECONDS, '', false );
```
These values are literally the default that we had set on the feature branch. Except that now they're options and are controllable via HTTP Response Headers. Also note that, if the options already exist on the database, we won't overwrite them here, too. :) (the priority comes from the data on the server, always)

fjorgemota commented on includes/promoted-jobs/class-wp-job-manager-promoted-jobs-status-handler.php:
```diff
@@ -102,4 +126,24 @@ private function get_site_feed_url() {
+			}
+		}
+	}
+
```
This is the logic to update the option according to the response headers. If the value is the magic `false`, we delete the option, otherwise we verify if it is an integer and, if it is, we update the value on the database

fjorgemota commented on includes/promoted-jobs/class-wp-job-manager-promoted-jobs-status-handler.php:
```diff
@@ -73,6 +68,33 @@ public function fetch_updates() {
+	 * @return int The current interval for fetching updates.
+	 */
+	private function get_current_interval() {
+		$option_name = wp_doing_cron() ? self::CRON_INTERVAL_OPTION_KEY : self::WEBHOOK_INTERVAL_OPTION_KEY;
```
I'm using `wp_doing_cron()` here to detect whether this is a CRON request or not and use the appropriate option key.

renatho commented on includes/promoted-jobs/class-wp-job-manager-promoted-jobs-status-handler.php:
```diff
@@ -73,6 +68,33 @@ public function fetch_updates() {
+			// If the interval is not set or is zero, we don't update.
+			return false;
+		}
+		$last_execution = get_option( self::LAST_EXECUTION_OPTION_KEY, 0 );
```
Not a big deal, but I think it's worth also casting it to `int`, similar to the `get_current_interval`.

fjorgemota committed:
Add new options for intervals to the Data Cleaner

To be deleted when uninstalling the plugin

fjorgemota committed:
Cast $last_execution to int

Co-authored-by: Renatho De Carli Rosa <renatho@automattic.com>

renatho reviewed (approved):
Looks good and works well! Thank you for implementing it!

fjorgemota committed:
Rename LAST_EXECUTION to LAST_CHECK_KEY

fjorgemota committed:
Set the LAST_CHECK option before redirecting to WPJMCOM

To protect a case where the promoted job status update is permanently disabled

fjorgemota committed:
Move logic to initialize default options for a static method in Status Handler

fjorgemota committed:
Simplify the logic of the update_interval method

And make it not remove the webhook interval if the value is false or 0

fjorgemota commented:
Hey @renatho, thanks for your review!

As discussed on Slack, I changed a few additional commits to fix some issues that I discovered, and also to improve the code a bit more. Hence, I'm requesting your review again.

For documentation purposes, the issue that I discovered is that this approach had two issues:

1. If the cron for updating the status ran while the user was still filling the form, it would delete the options of intervals for webhooks and crons, causing the code to effectively block the webhook/cron from fetching the feed
2. If WPJMCOM returned `false` or `0` for the interval of the webhook, it would effectively block the webhook from running (which isn't ideal)

The 1º issue made this PR similar to #2561 (in the sense that the job status might never be updated) and was fixed in 3039e61. The 2º issue was fixed in aa6ef7f.

Sorry for the trouble! :)

fjorgemota committed:
Always update LAST_CHECK option when initializing defaults

In case we decide to update WPJMCOM logic to be more strict and somehow the cron runs before the data there is updated

renatho commented on includes/promoted-jobs/class-wp-job-manager-promoted-jobs-status-handler.php:
```diff
@@ -102,4 +126,47 @@ private function get_site_feed_url() {
+	 */
+	public static function initialize_defaults() {
+		$webhook_interval = 5 * MINUTE_IN_SECONDS;
+		update_option( self::LAST_CHECK_OPTION_KEY, time() - $webhook_interval, '', false );
```
Thank you, Fernando! Just one final thing I noticed:
Since we're handling the webhook interval dynamically, I think we should get this interval from the option and use this only if the option doesn't exist, right?

The rest is looking great to me! I'll just wait it and the conflict fix before running a final test to approve. ;)

fjorgemota replied:
I implemented your suggestion in 4e0b5a1. Thanks!

fjorgemota committed:
Consider current webhook interval when updating LAST CHECK to not block webhook updates

fjorgemota committed:
Merge branch 'feature/promoted-jobs' into change/dynamic-status-update-interval

renatho reviewed (approved):
Looks good and works well! 🚀🚀🚀
